demo_cqm.py

```diff
@@ -0,0 +1,359 @@
+
+    Args:
+        - dist_mat: Numpy matrix providing distance between cities i and j.
+        - city_names: List of all airport codes.
+        - verbose: Print to command-line for user.
+
+    Returns:
+        - G: NetworkX weighted graph of cities with distances on edges.
+    """
+
+    if verbose:
+        print("\nConstructing map...\n")
+
+    G = nx.Graph()
+
+    num_cities = len(city_names)
+    for i in range(num_cities):
+        for j in range(i+1, num_cities):
+            G.add_edge(city_names[i], city_names[j], weight=dist_mat[i,j])
+
+    return G
+
+def draw_graph(G, city_names, city_lats, city_longs):
+    """Visualizes the city graph and saves as file.
+
+    Args:
+        - G: NetworkX weighted graph of cities with distances on edges.
+        - city_names: List of all airport codes.
+        - city_lats: List of all airport lat coordinates.
+        - city_longs: List of all airport long coordinates.
+
+    All city info lists have airports in the same order, i.e. airport city_names[i]
+    has latitude city_lats[i] and longitude city_longs[i].
+
+    Returns:
+        None. Saves visual as 'complete_network.png'.
+    """
+    
+    positions = {}
+    for i in range(len(city_names)):
+        positions[city_names[i]] = [-city_longs[i], city_lats[i]]
+
+    nx.draw(G, pos=positions, with_labels=True)
+    plt.savefig('complete_network.png')
+    plt.close()
+
+def build_cqm(W, C, n, p, a, verbose=True):
+    """Builds constrained quadratic model representing the optimization problem.
+
+    Args:
+        - W: Numpy matrix. Represents passenger demand. Normalized with total demand equal to 1.
+        - C: Numpy matrix. Represents airline leg cost.
+        - n: Int. Number of cities in play.
+        - p: Int. Number of hubs airports allowed.
+        - a: Float in [0.0, 1.0]. Discount allowed for hub-hub legs.
+        - verbose: Print to command-line for user.
+
+    Returns:
+        - cqm: ConstrainedQuadraticModel representing the optimization problem.
+    """
+
+    if verbose:
+        print("\nBuilding CQM...\n")
+
+    # Initialize the CQM object
+    cqm = ConstrainedQuadraticModel()
+
+    # Objective: Minimize cost. min c'x+x'Qx
+    # See reference paper for full explanation.
+    M = np.sum(W, axis=0)+np.sum(W, axis=1)
+    Q = a*np.kron(W,C)
+
+    linear = ((M*C.T).T).flatten()
+
+    q1 = np.tile(np.arange(n**2), 
+                (n**2,1)).flatten('F')
+    q2 = np.tile(np.arange(n**2), n**2)
+    q3 = Q.flatten()
+    var_order = [(i,j) for i in range(n) for j in range(n)]
+    
+    obj = BinaryQuadraticModel.from_numpy_vectors(linear=linear, 
+                                                            quadratic=(q1, q2, q3), 
+                                                            offset=0, 
+                                                            variable_order = var_order,
+                                                            vartype='BINARY')
```

Comment:
I find the index calculations (with `np.tile`) needed to use `from_numpy_vectors` to not be very intuitive.  Alternatively, you can use the original `linear` and `Q` arrays directly, and then relabel the variables.  Up to you.
```suggestion
    obj = BinaryQuadraticModel(linear, Q, 'BINARY')
    obj.relabel_variables({idx: (i,j) for idx, (i,j) in
                           enumerate((i,j) for i in range(n) for j in range(n))})
```

Reply:
I ran into a lot of run time issues when I didn't use numpy to build this model. For example, using the code as is takes 0.15592503547668457 sec, while the suggested fix takes 21.302162885665894 sec.

Slack thread for more context around other methods I tried: https://dwavesys.slack.com/archives/CBN3G5T60/p1649245297186319

Reply:
Can you double-check?  Something seems off.  I tried both versions of this code in both dimod 0.10.12 and 0.10.17.  For me, the change above was faster in both cases.  To be clear, I had in mind this fix more for clarity, but AFAICT it also performs better.

Timings below are for everything inside of `build_cqm`:
| CQM construction | Ocean version | Time for `build_cqm` (s) |
|-|-|-|
| `from_numpy_vectors` | 0.10.12 | 0.24 |
| `relabel_variables` | 0.10.12 | 0.16 |
| `from_numpy_vectors` | 0.10.17 | 0.091 |
| `relabel_variables` | 0.10.17 | 0.063 |

Reply:
I was doing it in the same function and resetting the objective. Maybe that was the issue? 

I just split it into two separate demos and with dimod 0.10.17 I'm seeing that both versions have similar run times, so no issue with the change now.